sched: dynamic event handlers registration

Huang-Wei opened the pull request:
#### What type of PR is this?

/kind feature
/sig scheduling

#### What this PR does / why we need it:

Kubernetes scheduler now registers event handlers dynamically, rather than hard-coding the events registration logic. Please see more details in [Design Doc](https://docs.google.com/document/d/1RRlWKiFD77NpJPBOfEpUDsqNxi43wbwExlLO-gh3kfY/edit#).

This PR consists of 3 parts:

- move clusterEventMap to Configurator, so the clusterEventMap can be used during event handlers registration
- dynamic event handlers registration for core API resources (existing integration test `TestServiceAffinityEnqueue` can ensure the test coverage)
- dynamic event handlers registration for custom resources (new integration test `TestCustomResourceEnqueue` added to ensure the test coverage)

#### Which issue(s) this PR fixes:

Part of #100347.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Scheduler now registers event handlers dynamically.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [Design Doc]: https://docs.google.com/document/d/1RRlWKiFD77NpJPBOfEpUDsqNxi43wbwExlLO-gh3kfY/edit#
(join sig-scheduling google group to view the doc)
```

ahg-g commented on test/integration/scheduler/queue_test.go:
```diff
@@ -148,3 +161,174 @@ func TestServiceAffinityEnqueue(t *testing.T) {
+		[]string{"--disable-admission-plugins=ServiceAccount,TaintNodesByCondition", "--runtime-config=api/all=true"},
+		testfwk.SharedEtcd(),
+	)
+	testCtx := &testutils.TestContext{}
```
importing an integration test pkg here is strange, why not have this test as an integration test?

Huang-Wei replied:
I don't follow that. It's an integration test.

ahg-g replied:
ah, sorry, I didn't notice, I thought it was a unit test :)

ahg-g commented on pkg/scheduler/scheduler.go:
```diff
@@ -281,10 +286,33 @@ func New(client clientset.Interface,
 	return sched, nil
 }
 
+func unionedGVKsExcludingNodePod(m map[framework.ClusterEvent]sets.String) map[framework.GVK]framework.ActionType {
```
I would not exclude anything in this function, instead get the union of all and add explicit case statements that does nothing for the kinds that we always want to handle.

Huang-Wei replied:
Done.

Huang-Wei commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -370,46 +329,104 @@ func addAllEventHandlers(
-			DeleteFunc: sched.onServiceDelete,
-		},
-	)
-
-	informerFactory.Storage().V1().StorageClasses().Informer().AddEventHandler(
-		cache.ResourceEventHandlerFuncs{
-			AddFunc: sched.onStorageClassAdd,
-		},
-	)
+			//
+			// PvAdd: Pods created when there are no PVs available will be stuck in
+			// unschedulable queue. But unbound PVs created for static provisioning and
+			// delay binding storage class are skipped in PV controller dynamic
+			// provisioning and binding process, will not trigger events to schedule pod
+			// again. So we need to move pods to active queue on PV add for this
+			// scenario.
+			//
+			// PvUpdate: Scheduler.bindVolumesWorker may fail to update assumed pod volume
+			// bindings due to conflicts if PVs are updated by PV controller or other
+			// parties, then scheduler will add pod back to unschedulable queue. We
+			// need to move pods to active queue on PV update for this scenario.
+			informerFactory.Core().V1().PersistentVolumes().Informer().AddEventHandler(
+				buildEvtResHandler(at, framework.PersistentVolume, "Pv"),
+			)
+		case framework.PersistentVolumeClaim:
+			// MaxPDVolumeCountPredicate: add/update PVC will affect counts of PV when it is bound.
+			informerFactory.Core().V1().PersistentVolumeClaims().Informer().AddEventHandler(
+				buildEvtResHandler(at, framework.PersistentVolumeClaim, "Pvc"),
+			)
+		case framework.StorageClass:
+			if at&framework.Add != 0 {
+				informerFactory.Storage().V1().StorageClasses().Informer().AddEventHandler(
+					cache.ResourceEventHandlerFuncs{
+						AddFunc: sched.onStorageClassAdd,
+					},
+				)
+			}
```
> do we really think that a user will configure a scheduler without registering the volume-based plugins?

Nope, I think volume-based plugins are always needed. The design intention is that they're not interacting with internal cache (but pod and node do), so they fit the `switch` case naturally, to leverage `buildEvtResHandler` generaizing their event handler instantiation.

ahg-g replied:
We could have `buildEvtResHandler` function moved outside, so we can still reuse it. The way I am interpreting this switch statement is whether or not those events are necessary to register. If we are saying the storage based ones should always be registered, then I think it is better to treat them like pod/node.

Huang-Wei replied:
(missed this comment earlier)

Another benefit is that we can generalize the registration for add/update/delete events dynamically - for example, currently storage-related registrations are hard-coded, some register add/update, some register add, etc. But if we leverage the unioned cluster events, we can dynamically register the event handlers - which will be resilient to adapt to further code changes such as add/delete events. WDYT?

Huang-Wei commented on pkg/scheduler/scheduler.go:
```diff
@@ -281,10 +286,33 @@ func New(client clientset.Interface,
+	// options.kubeConfig can be nil in tests.
+	if options.kubeConfig != nil {
+		dynClient := dynamic.NewForConfigOrDie(options.kubeConfig)
+		dynInformerFactory = dynamicinformer.NewFilteredDynamicSharedInformerFactory(dynClient, 0, v1.NamespaceAll, nil)
```
In terms of default scheduler, there is no plugin registering CR events, so IMO instantiating a `dynInformerFactory` won't do any concrete work.

Huang-Wei replied:
yes, just like coreInformerFactory, by default the embedded informers are empty.

ahg-g replied:
hmm, so someone needs to initiate an informer then, so that makes it even more necessary to pass this factory via framework handler to the plugins that will use it, right?

Huang-Wei replied:
Yes, that's your point at https://github.com/kubernetes/kubernetes/pull/101394#discussion_r620264539. For now, plugin developers instantiate **typed** informer and register event handlers accordingly.

ahg-g replied:
It is related, but might be different. If you remember, an informer is initiated only if it is explicitly accessed before starting the the informer. Is this the case here (I don't have experience with dynamic informer, so asking just to make sure)?

Huang-Wei replied:
> If you remember, an informer is initiated only if it is explicitly accessed before starting the the informer.

This is true. However, the framework doesn't really care about the cache for CRs - the out-of-tree plugins are responsible to build their own cache, if needed. The hook setup here is fairly simple: just to guarantee that upon some CR events, partial unschedulable Pods can be moved properly.

(BTW: I rethink the necessity of offering the dynInformerFactory in framework handle. See my other comment https://github.com/kubernetes/kubernetes/pull/101394#discussion_r620773413)

Huang-Wei commented on pkg/scheduler/scheduler.go:
```diff
@@ -281,10 +286,33 @@ func New(client clientset.Interface,
 
-	addAllEventHandlers(sched, informerFactory)
+	// Build dynamic client and dynamic informer factory
+	var dynInformerFactory dynamicinformer.DynamicSharedInformerFactory
```
Good point. I need to think about it. I'm a bit concerned that would add extra burden for plugin developers, such as fetching the GVR before add event handlers, etc. Probably need to compare it with how controller-runtime obstacles similar problems, and come up with a better abstraction.

Huang-Wei replied:
I rethink it a bit. It's better not to enforce out-of-tree plugins to use dynamic informers as typed ones are more applicable and easy to manipulate, in both core logic and tests.

ahg-g replied:
We are not forcing them to use it, but I think it is useful to make it available to them since the cache is already populated for the types they registered events for. If they create there own informer for the same type, aren't we wasting memory because we will have two informerFactor caches?

Huang-Wei replied:
> but I think it is useful to make it available to them since the cache is already populated for the types they registered events for

Not quite sure the plugin developers would really use it, as the typed informer and other generated stuff are way easier to manipulate.

> If they create their own informer for the same type, aren't we wasting memory because we will have two informerFactory caches?

Yes, this is true.

Huang-Wei commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -370,46 +329,104 @@ func addAllEventHandlers(
+			// Fall back to try dynamic informers.
+			gvr, _ := schema.ParseResourceArg(string(gvk))
+			dynInformer := dynInformerFactory.ForResource(*gvr).Informer()
+			// go func(stopCh <-chan struct{}) {
+			// 	dynInformer.AddEventHandler(
+			// 		buildEvtResHandler(at, gvk, strings.Title(gvr.Resource)),
+			// 	)
+			// 	dynInformer.Run(stopCh)
+			// }(sched.StopEverything)
```
done.

Huang-Wei commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -370,46 +329,104 @@ func addAllEventHandlers(
+			}
+			// GVK is expected to be at least 3-folded, separated by dots.
+			// <kind in plural>.<version>.<group>
+			// Legal examples:
```
done.

Huang-Wei commented on pkg/scheduler/eventhandlers.go:
```diff
@@ -370,46 +329,104 @@ func addAllEventHandlers(
+			// Legal examples:
+			// - foos.v1.example.com
+			// - bars.v1beta1.a.b.c
+			// Illegal examples:
```
done.

Huang-Wei commented:
@ahg-g any further comments on this PR?

ahg-g commented:
/lgtm
/hold 

for squash

Huang-Wei committed:
Scheduler now registers event handlers dynamically

- move clusterEventMap to Configurator
- dynamic event handlers registration for core API resources
- dynamic event handlers registration for custom resources

Huang-Wei commented:
@ahg-g Commits squashed.

Huang-Wei commented:
@dims Could you take a look at the vendors change, or cascade to the correct reviewer? Thanks.

dims commented:
/approve

the `vendor/ ` changes are good.

Huang-Wei commented:
/retest

ahg-g commented:
/lgtm

Huang-Wei commented:
/hold cancel
/retest

Huang-Wei commented:
/retest
